src/main/java/net/dries007/tfc/objects/te/TELamp.java

```diff
@@ -0,0 +1,127 @@
+
+import net.dries007.tfc.ConfigTFC;
+import net.dries007.tfc.objects.fluids.capability.FluidHandlerSided;
+import net.dries007.tfc.objects.fluids.capability.FluidTankCallback;
+import net.dries007.tfc.objects.fluids.capability.IFluidHandlerSidedCallback;
+import net.dries007.tfc.objects.fluids.capability.IFluidTankCallback;
+import net.dries007.tfc.objects.items.itemblock.ItemBlockMetalLamp;
+
+public class TELamp extends TETickCounter implements IFluidTankCallback, IFluidHandlerSidedCallback
+{
+    public final static int CAPACITY = ConfigTFC.GENERAL.metalLampCapacity;
+
+    private final FluidTank tank = new FluidTankCallback(this, 0, CAPACITY);
+    private boolean powered = false;
+
+    public TELamp()
+    {
+        super();
+        this.tank.setTileEntity(this);
+    }
+
+    @Override
+    public void setAndUpdateFluidTank(int fluidTankID)
+    {
+        IBlockState state = world.getBlockState(pos);
+        world.notifyBlockUpdate(pos, state, state, 3);
+    }
+
+    @Override
+    public boolean hasCapability(Capability<?> capability, @Nullable EnumFacing facing)
+    {
+        return capability == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T getCapability(Capability<T> capability, @Nullable EnumFacing facing)
+    {
+        if (capability == CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY)
+        {
+            return (T) new FluidHandlerSided(this, tank, facing);
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean canFill(FluidStack resource, EnumFacing side)
+    {
+        return ItemBlockMetalLamp.getValidFluids().contains(resource.getFluid());
+    }
+
+    @Override
+    public boolean canDrain(EnumFacing side)
+    {
+        return true;
+    }
+
+    public void onBreakBlock(World worldIn, BlockPos pos, IBlockState state)
+    {
+        InventoryHelper.spawnItemStack(worldIn, pos.getX(), pos.getY(), pos.getZ(), getItemStack(this, state));
+    }
+
+    public ItemStack getItemStack(TELamp tel, IBlockState state)
+    {
+        ItemStack stack = new ItemStack(state.getBlock());
+        IFluidHandlerItem itemCap = stack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null);
+        IFluidHandler teCap = tel.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, null);
+        if (itemCap != null && teCap != null)
+        {
+            itemCap.fill(teCap.drain(CAPACITY, false), true); //don't drain creative item
+        }
+        return stack;
+    }
+
+    @Override
+    public void readFromNBT(NBTTagCompound nbt)
+    {
+        super.readFromNBT(nbt);
+        tank.readFromNBT(nbt.getCompoundTag("tank"));
+        if (tank.getFluidAmount() > tank.getCapacity())
+        {
+            // Fix config changes
+            FluidStack fluidStack = tank.getFluid();
+            //noinspection ConstantConditions
```

Comment:
This can be null, don't ignore this, null check instead.

Reply:
As with some of these, you're finding problems with code I copied verbatim from other places in TFC: TEBarrel#readFromItemTag and #readFromNBT
That's fun.

Reply:
Fixed here and in the two places in TEBarrel